unit_tests/string_buffer/src/test_string_buffer.cpp

```diff
@@ -0,0 +1,189 @@
+};
+
+class test_string_buffer
+{
+private:
+  char m_buf[sizeof (prefix) + N + sizeof (suffix)];// working buffer
+  size_t m_dim;                                     //_ref[_dim]
+  size_t m_len;                                     // sizeof(_ref)
+  char* m_ref;                                      // reference buffer
+  allocator::stack m_stack_allocator;
+  allocator::affix<allocator::stack, prefix, suffix> m_affix_allocator;
+  allocator::blk m_blk;
+  string_buffer m_sb;
+
+public:
+  test_string_buffer ()
+    : m_buf ()
+    , m_dim (1024)
+    , m_len (0)
+    , m_ref ((char*)calloc (m_dim, 1))
+    , m_stack_allocator (m_buf, sizeof (m_buf))
+    , m_affix_allocator (m_stack_allocator)
+    , m_blk ()
+    , m_sb ()
+  {
+  }
+
+  void operator() (size_t size)
+  {// prepare for a test with a buffer of <size> bytes
+    if (m_dim < size)
+      {// adjust internal buffer is necessary
+        do
+          {
+            m_dim *= 2;
+          }
+        while (m_dim < size);
+        char* p = (char*)malloc (m_dim);
+      }
+    m_len = 0;
+    m_stack_allocator.~stack ();
+    m_blk = m_affix_allocator.allocate (size);
+    m_sb.set (size, m_blk.ptr);
+  }
+
+  template<size_t Size, typename... Args>
+  void operator() (const char* file, int line, const char (&format)[Size], Args&&... args)
+  {
+    int len = snprintf (m_ref + m_len, m_len < m_blk.dim ? m_blk.dim - m_len : 0, format, args...);
+    if (len < 0)
+      {
+        ERR ("[%s(%d)] StrBuf([%zu]) snprintf()=%d", file, line, m_blk.dim, len);
+        return;
+      }
+    else
+      {
+        m_len += len;
+      }
+    m_sb (format, args...);
+    if (m_sb.len () != m_len)
+      {
+        ERR ("[%s(%d)] StrBuf([%zu]) len()=%zu expect %zu", file, line, m_blk.dim, m_sb.len (), m_len);
+        return;
+      }
+    if (strcmp (m_sb, m_ref))
+      {
+        ERR ("[%s(%d)] StrBuf([%zu]) {\"%s\"} expect{\"%s\"}", file, line, m_blk.dim, (const char*)m_sb, m_ref);
+        return;
+      }
+    if (m_affix_allocator.check (m_blk))
+      {
+        ERR ("[%s(%d)] StrBuf(buf[%zu]) memory corruption", file, line, m_blk.dim);
+        return;
+      }
+  }
+
+  void operator() (const char* file, int line, char ch)
+  {
+    if (m_len + 1 < m_blk.dim)
+      {// include also '\0'
+        m_ref[m_len]     = ch;
+        m_ref[m_len + 1] = '\0';
+      }
+    ++m_len;
+
+    m_sb += ch;
+    if (strcmp ((const char*)m_sb, m_ref))
+      {
+        ERR ("[%s(%d)] StrBuf([%zu]) {\"%s\"} expect {\"%s\"}", file, line, m_blk.dim, (const char*)m_sb, m_ref);
+        return;
+      }
+    if (m_affix_allocator.check (m_blk))
+      {
+        ERR ("[%s(%d)] StrBuf([%zu]) memory corruption", file, line, m_blk.dim);
+      }
+  }
+};
+test_string_buffer test;
+
+#define SB_FORMAT(format, ...) test (__FILE__, __LINE__, format, ##__VA_ARGS__)
+#define SB_CHAR(ch) test (__FILE__, __LINE__, ch);
+
+int main (int argc, char** argv)
+{
+  enum Flags
```

Comment:
Standard naming for enums: lowercase enum name, uppercase enum values.
enum flags
{
  FL_DEBUG = (1 << 0),
  FL_DEBUG = (1 << 1)
};